include/ginkgo/core/distributed/vector.hpp

```diff
@@ -0,0 +1,416 @@
+     * is complex, the result is created by viewing the complex vector with as
+     * real with a reinterpret_cast with twice the number of columns and
+     * double the stride.
+     */
+    std::unique_ptr<const real_type> create_real_view() const;
+
+protected:
+    /**
+     * Creates an empty distributed vector with a specified size
+     *
+     * @param exec  Executor associated with vector
+     * @param comm  Communicator associated with vector, the default is
+     *              MPI_COMM_WORLD
+     * @param partition  Partition of global rows
+     * @param global_size  Global size of the vector
+     * @param local_size  Processor-local size of the vector
+     * @param stride  Stride of the local vector.
+     */
+    Vector(std::shared_ptr<const Executor> exec, mpi::communicator comm,
+           dim<2> global_size, dim<2> local_size, size_type stride);
+
+    /**
+     * Creates an empty distributed vector with a specified size
+     *
+     * @param exec  Executor associated with vector
+     * @param comm  Communicator associated with vector, the default is
+     *              MPI_COMM_WORLD
+     * @param partition  Partition of global rows
+     * @param global_size  Global size of the vector
+     * @param local_size  Processor-local size of the vector, uses local_size[1]
+     *                    as the stride
+     */
+    explicit Vector(std::shared_ptr<const Executor> exec,
+                    mpi::communicator comm = mpi::communicator(MPI_COMM_WORLD),
+                    dim<2> global_size = {}, dim<2> local_size = {});
+
+    /**
+     * Creates a distributed vector from local vectors with a specified size.
+     *
+     * @param exec  Executor associated with this vector
+     * @param comm  Communicator associated with this vector
+     * @param global_size  The global size of the vector
+     * @param local_vector  The underlying local vector, the data will be moved
+     *                      into this
+     */
+    Vector(std::shared_ptr<const Executor> exec, mpi::communicator comm,
+           dim<2> global_size, local_vector_type* local_vector);
+
+    /**
+     * Creates a distributed vector from local vectors. The global size will
+     * be deduced from the local sizes, which will incur a collective
+     * communication.
+     *
+     * @param exec  Executor associated with this vector
+     * @param comm  Communicator associated with this vector
+     * @param local_vector  The underlying local vector, the data will be moved
+     *                      into this
+     */
+    Vector(std::shared_ptr<const Executor> exec, mpi::communicator comm,
+           local_vector_type* local_vector);
+
+    void resize(dim<2> global_size, dim<2> local_size);
+
+    void apply_impl(const LinOp*, LinOp*) const override;
+
+    void apply_impl(const LinOp*, const LinOp*, const LinOp*,
+                    LinOp*) const override;
+
+private:
+    local_vector_type local_;
```

Comment:
use Dense directly not with shared pointer?

Reply:
value semantics ;)